Khan/webpack
# Speedup source map plugin

**What kind of change does this PR introduce?**

This PR adds a new feature.  In particular it adds a `skipExistingFiles` option to `SourceMapDevToolPlugin` which skips generating source maps for bundles whose source map files already exist on disk.  This setting should be used with `DisableEmitPlugin`.  This setting is not compatible with inline source maps.

**Did you add tests for your changes?**

Not yet.

**Does this PR introduce a breaking change?**

It should not because the new behavior is behind a new configuration option to SourceMapDevToolPlugin.

**What needs to be documented once your changes are merged?**

The only thing that needs documentation is the new `skipExistingFiles` option.

**Test Plan**

- Update webapp's wepback.prod.config to replace the `devtool` setting with an explicit `SourceMapDevToolPlugin` like so:
```
// Generate sourcemaps.
            // See https://webpack.js.org/plugins/source-map-dev-tool-plugin/.
            new webpack.SourceMapDevToolPlugin({
                module: true,
                // Quicker to generate but doesn't allow setting breakpoints
                // on columns or stepping through statements within the same
                // line.
                columns: false,
                // Simplify and speed up source mapping by using line to line
                // source mappings for matched modules.
                lineToLine: true,
                noSources: false,
                // Generates discrete files containing the source maps.
                // NOTE: currently we only support replacing "[file]".
                filename: "[file].map",
                // This setting is ignored if `filename` is undefined.
                skipExistingFiles: true,
            }),
```
- run `yarn build:webpack-prod-profile --no-i18n` in webapp
- see `36ms after chunk asset optimization` in the output
- run `yarn test` in webpack and see that all the tests pass
<img width="985" alt="screen shot 2019-01-18 at 5 54 12 pm" src="https://user-images.githubusercontent.com/1044413/51418687-f8fb5580-1b52-11e9-99d9-0b40d8273bd2.png">
